src/features/class-full-page-cache-404.php

```diff
@@ -0,0 +1,337 @@
+		 */
+		if ( ! wp_using_ext_object_cache() || ! is_ssl() ) {
+			return;
+		}
+
+		// Return 404 page cache on template_redirect.
+		add_action( 'template_redirect', [ self::class, 'action__template_redirect' ], 1 );
+
+		// For the Guaranteed 404 page, hook in on WP to start output buffering, to capture the HTML.
+		add_action( 'wp', [ self::class, 'action__wp' ] );
+
+		// Replenish the cache every hour.
+		if ( ! wp_next_scheduled( 'wp_404_cache' ) ) {
+			wp_schedule_event( time(), 'hourly', 'wp_404_cache' );
+		}
+
+		// Callback for Cron Event.
+		add_action( 'wp_404_cache', [ self::class, 'trigger_404_page_cache' ] );
+		add_action( 'wp_404_cache_single', [ self::class, 'trigger_404_page_cache' ] );
+	}
+
+	/**
+	 * Get 404 Page Cache and return early if found.
+	 */
+	public static function action__template_redirect(): void {
+
+		// Don't cache if user is logged in.
+		if ( is_user_logged_in() ) {
+			return;
+		}
+
+		// Don't cache if not a 404.
+		if ( ! is_404() ) {
+			return;
+		}
+
+		// Allow this URL through, as this request will populate the cache.
+		if ( isset( $_SERVER['REQUEST_URI'] ) && self::TEMPLATE_GENERATOR_URI === $_SERVER['REQUEST_URI'] ) {
+			return;
+		}
+
+		echo self::get_cached_response_with_headers(); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+
+		// If we're testing, don't exit, die instead.
+		if ( defined( 'MANTLE_IS_TESTING' ) && MANTLE_IS_TESTING ) {
+			wp_die( '', '', [ 'response' => 404 ] );
+		}
+
+		exit;
+	}
+
+	/**
+	 * Get cached response with headers.
+	 *
+	 * @return string
+	 */
+	public static function get_cached_response_with_headers(): string {
+		$stale_cache_in_use = false;
+		$cache              = self::get_cache();
+
+		if ( false === $cache ) {
+			$cache              = self::get_stale_cache();
+			$stale_cache_in_use = true;
+		}
+
+		if ( ! empty( $cache ) && is_string( $cache ) ) {
+			$html = self::prepare_response( $cache );
+
+			self::send_header( 'HIT', $stale_cache_in_use );
+
+			// Cached content is already escaped.
+			return $html; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+		}
+
+		// Schedule a single event to generate the cache immediately.
+		if ( ! wp_next_scheduled( 'wp_404_cache_single' ) ) {
+			wp_schedule_single_event( time(), 'wp_404_cache_single' );
+		}
+
+		self::send_header( 'MISS' );
+
+		// If no cache, return an empty string.
+		return '';
+	}
+
+	/**
+	 * Send X-WP-404-Cache HTTP Header.
+	 *
+	 * @param string $type HIT or MISS.
+	 * @param bool   $stale Whether the stale cache is in use. Default false.
+	 */
+	public static function send_header( string $type, bool $stale = false ): void {
+
+		if ( headers_sent() ) {
+			return;
+		}
+
+		if ( ! $stale && 'HIT' === $type ) {
+			header( 'X-WP-404-Cache: HIT' );
+		} elseif ( $stale && 'HIT' === $type ) {
+			header( 'X-WP-404-Cache: HIT (stale)' );
+		} elseif ( 'MISS' === $type ) {
+			header( 'X-WP-404-Cache: MISS' );
+		}
```

Comment:
Personal preference / 6 of one, half-dozen of the other

```suggestion
		if ( 'HIT' === $type ) {
			if ( $stale ) {
				header( 'X-WP-404-Cache: HIT (stale)' );
			} else {
				header( 'X-WP-404-Cache: HIT' );
			}
		} elseif ( 'MISS' === $type ) {
			header( 'X-WP-404-Cache: MISS' );
		}
```